nemo/collections/asr/parts/submodules/jasper.py

```diff
@@ -462,72 +432,66 @@ def __init__(
-            y = self._se_pool_step(x, mask)  # [B, C, 1]
-            y = y.transpose(1, -1)  # [B, 1, C]
-            y = self.fc(y)  # [B, 1, C]
-            y = y.transpose(1, -1)  # [B, C, 1]
+            # masked_fill needs to be here because of TRT issue with Where()
+            x = x.float().masked_fill(mask, 0.0)  # mask padded values explicitly to 0
+            y = self._se_pool_step(x, mask).to(x.dtype)
+
+        y = y.transpose(1, -1)  # [B, 1, C]
+        y = self.fc(y)  # [B, 1, C]            
+        y = y.transpose(1, -1)  # [B, C, 1]
 
         # Note: Keep for future, in case we improve WER from doing so.
         # if self.context_window >= 0:
         #     y = F.interpolate(y, size=x.shape[-1], mode=self.interpolation_mode)
 
         y = torch.sigmoid(y)
         y = x * y
-        return y, lengths
 
+        return y, lengths
+    
     def _se_pool_step(self, x, mask):
         # Negate mask back to represent 1 for signal and 0 for padded timestep.
         mask = ~mask
-
-        if self.context_window < 0:
+        if self.context_window < 0 or not self.training:
             # [B, C, 1] - Masked Average over value + padding.
-            y = torch.sum(x, dim=-1, keepdim=True) / mask.sum(dim=-1, keepdim=True).type(x.dtype)
+            y = torch.sum(x, dim=-1, keepdim=True).div(mask.sum(dim=-1, keepdim=True).to(x.dtype))
         else:
             # [B, C, 1] - Masked Average over value + padding with limited context.
             # During training randomly subsegments a context_window chunk of timesteps.
             # During inference selects only the first context_window chunk of timesteps.
-            if self.training:
-                y = _se_pool_step_script_train(x, self.context_window, mask)
-            else:
-                y = _se_pool_step_script_infer(x, self.context_window, mask)
+            y = _se_pool_step_script_train(x, self.context_window, mask)
         return y
 
     def set_max_len(self, max_len):
         """ Sets maximum input length.
             Pre-calculates internal seq_range mask.
         """
+        if (self.max_len > max_len):
+            return
         self.max_len = max_len
         device = next(self.parameters()).device
         seq_range = torch.arange(0, self.max_len, device=device)
-        self.register_buffer('seq_range', seq_range, persistent=False)
+        if hasattr(self, 'seq_range'):
+            self.seq_range = seq_range
+        else:
+            self.register_buffer('seq_range', seq_range, persistent=False)
 
-    def make_pad_mask(self, seq_lens, max_audio_length, device=None):
+    def make_pad_mask(self, seq_lens, max_audio_length):
         """Make masking for padding."""
-        if device and self.seq_range.device != device:
-            self.seq_range = self.seq_range.to(device)
-
-        if self.seq_range.device != seq_lens.device:
-            seq_lens = seq_lens.to(self.seq_range.device)
```

Comment:
These two lines are necessary for inference without Pytorch Lightning - there is no guarantee that the tensor are on correct device when you call model.transcribe() or other high level apis that don't depend on PTL

Reply:
^